patterns/xilinx_bootgen.hexpat

```diff
@@ -0,0 +1,370 @@
+        PpkKeyType ppk_key_type : 1;
+        AcVersion authentication_certificate_version : 2;
+        AcFormat authentication_certificate_format : 2;
+        PpkKeySelect ppk_key_select : 2;
+        SpkUserEfuseSelect spk_user_efuse_select : 2;
+        padding : 12;
+    };
+
+    struct AuthenticationCertificates {
+        AuthenticationCertificationHeader authentication_header;
+        u32 spk_id;
+        u32 udf[14];
+        u32 ppk_mod[128];
+        u32 ppk_mod_ext[128];
+        u32 ppk_exponent;
+        padding[60];
+        u32 spk_mod[128];
+        u32 spk_mod_ext[128];
+        u32 spk_exponent;
+        padding[60];
+        u32 spk_signature[128];
+        u32 boot_header_signature[128];
+        u32 partition_signature[128];
+    };
+
+    enum VectorLocation : u8 {
+        Low = 0,
+        High = 1,
+    };
+
+    enum EarlyHandoff : u8 {
+        Disabled = 0,
+        Enabled = 1,
+    };
+
+    enum Endianness : u8 {
+        Little = 0,
+        Big = 1,
+    };
+
+    enum PartitionOwner : u8 {
+        Fsbl = 0,
+        UBoot = 1,
+    };
+
+    enum RsaCertPresent : u8 {
+        No = 0,
+        Yes = 1,
+    };
+
+    enum ChecksumType : u8 {
+        None = 0,
+        Sha3 = 3,
+    };
+
+    enum DestCpu : u8 {
+        None = 0,
+        A53_0 = 1,
+        A53_1 = 2,
+        A53_2 = 3,
+        A53_3 = 4,
+        R5_0 = 5,
+        R5_1 = 6,
+        R5Lockstep = 7,
+        Pmu = 8,
+    };
+
+    enum EncryptionPresent : u8 {
+        No = 0,
+        Yes = 1,
+    };
+
+    enum DestDevice : u8 {
+        None = 0,
+        Ps = 1,
+        Pl = 2,
+    };
+
+    enum A5xExecState : u8 {
+        Aarch64 = 0,
+        Aarch32 = 1,
+    };
+
+    enum ExceptionLevel : u8 {
+        El0 = 0,
+        El1 = 1,
+        El2 = 2,
+        El3 = 3,
+    };
+
+    enum TrustZone : u8 {
+        NonSecure = 0,
+        Secure = 1,
+    };
+
+    bitfield PartitionAttributeBits {
+        TrustZone trustzone : 1;
+        ExceptionLevel exception_level : 2;
+        A5xExecState a5x_exec_state : 1;
+        DestDevice destination_device : 3;
+        EncryptionPresent encryption_present : 1;
+        DestCpu destination_cpu : 4;
+        ChecksumType checksum_type : 3;
+        RsaCertPresent rsa_authentication_certificate_present : 1;
+        PartitionOwner partition_owner : 2;
+        Endianness endianness : 1;
+        EarlyHandoff early_handoff : 1;
+        padding : 3;
+        VectorLocation vector_location : 1;
+        padding : 8;
+    };
+
+    struct PartitionHeader {
+        u32 encrypted_partition_data_word_length;
+        u32 unencrypted_data_word_length;
+        u32 total_partition_word_length;
+
+        NullableWordPtr<PartitionHeader, u32> next_partition_header_offset;
+
+        u64 destination_execution_address;
+        u64 destination_load_address;
+        NullableWordPtr<std::Array<u8,parent.encrypted_partition_data_word_length*4>, u32> actual_partition_word_offset;
```

Comment:
This is an oddity I ran into with the templating system, I must refer to the array size by `parent.x` for this to work. I'm guessing the parser is evaluating the template expression from inside NullableWordPtr.